include/deal.II/base/tensor.h

```diff
@@ -16,14 +16,259 @@
+ * vectors, matrices, etc, depending on the values of the template arguments
+ * @p dim and @p spacedim. We therefore need a class that acts as a scalar
+ * (i.e. @p Number) for all purposes but is part of the Tensor template
+ * family.
+ *
+ * @tparam dim An integer that denotes the dimension of the space in which
+ * this tensor operates. This of course equals the number of coordinates that
+ * identify a point and rank-1 tensor. Since the current object is a rank-0
+ * tensor (a scalar), this template argument has no meaning for this class.
+ *
+ * @tparam Number The data type in which the tensor elements are to be stored.
+ * This will, in almost all cases, simply be the default @p double, but there
+ * are cases where one may want to store elements in a different (and always
+ * scalar) type. It can be used to base tensors on @p float or @p complex
+ * numbers or any other data type that implements basic arithmetic operations.
+ * Another example would be a type that allows for Automatic Differentiation
+ * (see, for example, the Sacado type used in step-33) and thereby can
+ * generate analytic (spatial) derivatives of a function that takes a tensor
+ * as argument.
+ *
+ * @ingroup geomprimitives
+ * @author Wolfgang Bangerth, 2009, Matthias Maier, 2015
+ */
+template <int dim, typename Number>
+class Tensor<0,dim,Number>
+{
+public:
+  /**
+   * Provide a way to get the dimension of an object without explicit
+   * knowledge of it's data type. Implementation is this way instead of
+   * providing a function <tt>dimension()</tt> because now it is possible to
+   * get the dimension at compile time without the expansion and preevaluation
+   * of an inlined function; the compiler may therefore produce more efficient
+   * code and you may use this value to declare other data types.
+   */
+  static const unsigned int dimension = dim;
+
+  /**
+   * Publish the rank of this tensor to the outside world.
+   */
+  static const unsigned int rank = 0;
+
+  /**
+   * Number of independent components of a tensor of rank 0.
+   */
+  static const unsigned int n_independent_components = 1;
+
+  /**
+   * Declare a type that has holds real-valued numbers with the same precision
+   * as the template argument to this class. For std::complex<number>, this
+   * corresponds to type number, and it is equal to Number for all other
+   * cases. See also the respective field in Vector<Number>.
+   *
+   * This typedef is used to represent the return type of norms.
+   */
+  typedef typename numbers::NumberTraits<Number>::real_type real_type;
+
+  /**
+   * Type of objects encapsulated by this container and returned by
+   * operator[](). This is a scalar number type for a rank 0 tensor.
+   */
+  typedef Number value_type;
+
+  /**
+   * Declare an array type which can be used to initialize an object of this
+   * type statically. In case of a a tensor of rank 0 this is just the scalar
+   * number type Number.
+   */
+  typedef Number array_type;
```

Comment:
> out of curiosity, would everything continue to work if you had 
> ```
>   typedef Number array_type[1];
> ```

It would require to add another constructor and an assignment operator
for the plain Number type to Tensor<0,dim,Number>.